junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java

```diff
@@ -1264,6 +1293,82 @@ public void close() throws IOException {
 
 	}
 
+	static class FactoryWithAnnotatedElementNameAsPrefixTestCase {
+
+		@TempDir(factory = Factory.class)
+		private Path tempDir1;
+
+		private Path tempDir2;
+
+		@BeforeEach
+		void setUp(@TempDir(factory = Factory.class) Path tempDir2) {
+			this.tempDir2 = tempDir2;
+		}
+
+		@Test
+		void test(@TempDir(factory = Factory.class) Path tempDir3) {
+			assertThat(tempDir1.getFileName()).asString().startsWith("tempDir1");
+			assertThat(tempDir2.getFileName()).asString().startsWith("tempDir2");
+			assertThat(tempDir3.getFileName()).asString().startsWith("tempDir3");
+		}
+
+		private static class Factory implements TempDirFactory {
+
+			@Override
+			public Path createTempDirectory(AnnotatedElementContext elementContext, ExtensionContext extensionContext)
+					throws Exception {
+				return Files.createTempDirectory(getName(elementContext.getAnnotatedElement()));
+			}
+
+			private static String getName(AnnotatedElement element) {
+				return element instanceof Field ? ((Field) element).getName() : ((Parameter) element).getName();
+			}
+
+		}
+
+	}
+
+	static class FactoryWithCustomAnnotationTestCase {
+
+		@Prefix("field")
+		@TempDir(factory = Factory.class)
+		private Path tempDir1;
+
+		private Path tempDir2;
+
+		@BeforeEach
+		void beforeEach(@Prefix("beforeEach") @TempDir(factory = Factory.class) Path tempDir2) {
+			this.tempDir2 = tempDir2;
+		}
+
+		@Test
+		void test(@Prefix("method") @TempDir(factory = Factory.class) Path tempDir3) {
+			assertThat(tempDir1.getFileName()).asString().startsWith("field");
+			assertThat(tempDir2.getFileName()).asString().startsWith("beforeEach");
+			assertThat(tempDir3.getFileName()).asString().startsWith("method");
+		}
+
+		@Target({ ANNOTATION_TYPE, FIELD, PARAMETER })
+		@Retention(RUNTIME)
+		private @interface Prefix {
```

Comment:
```suggestion
		@Retention(RUNTIME)
		@TempDir(factory = Factory.class)
		private @interface CustomTempDir {
```

Reply:
Perhaps it might make sense to move the tests from `TempDirectoryMetaAnnotationTests` to `TempDirectoryPerDeclarationTests` and improve them.

Reply:
Eventually, I decided to keep them separate as `TempDirectoryMetaAnnotationTests` have no relation to factories.